### ISSUE TITLE
mkldnn batch norm maybe diff with mkl batch norm

### ISSUE THREAD

**Issue by luotao1** [PaddlePaddle/Paddle#13646]
The test machine is: E5-2620 v2, which only have AVX Instruction set.

When I inference resnet50 model: https://github.com/PaddlePaddle/Paddle/compare/develop...luotao1:resnet50?expand=1

The commit id is: https://github.com/PaddlePaddle/Paddle/commit/643b6faa0ced3304d5531dfa6a87a7fb39a54e0f

I run `ctest -R test_analyzer_resnet50 -V`.  following Test fails:
```
// Compare result of NativeConfig and AnalysisConfig
TEST(Analyzer_resnet50, compare) 
```
The error is that there is > 0.001 diff for all result data.
```
95: The difference between pdata_ref[j] and pdata[j] is 0.00164794921875, which exceeds 1e-3, where
95: pdata_ref[j] evaluates to 451.11993408203125,
95: pdata[j] evaluates to 451.1182861328125, and
95: 1e-3 evaluates to 0.001.
```
When `FLAGS_with_mkldnn=true`:
- NativeConfig: mkl
- AnalysisConfig: mkldnn. And to avoid `fc_fuse_pass`, I exclude this `fc_fuse_pass` as well. Thus, the only diff is between mkl op and mkldnn op.

However, as `test_analyzer_ocr` runs sucessfully. We compare the ops in [ocr model](https://github.com/PaddlePaddle/Paddle/issues/13574#issuecomment-424638054) and resnet50 model.  **The difference is batch_norm op**, which doesn't exist in ocr model.

Besides I also merge #13486 to see whether `conv bn fuse pass` runs sucessfully.
But both I set `cfg->_use_mkldnn = true` and `cfg->_use_mkldnn = false`, there is still diff.

Could you help see it?

**Comment by Sand3r-** [comment]
Thank you for the detailed report! We're on our way to investigate the problem with @sfraczek.

**Comment by mrysztow** [comment]
@luotao1 can you say, what are severity and priority of this request? We don't have any v2 platforms to start immediate investigations and if it is not "nice to have" only, need to obtain proper platform

**Comment by luotao1** [comment]
> what are severity and priority of this request?

This PR has high priority, since the majority of online machines are V2 and V4 platforms, and this resnet50 model will be used online. 

> We don't have any v2 platforms to start immediate investigations

Does this PR relate with the platform but not codes? Could it run successfully on 6148 platform?

**Comment by mrysztow** [comment]
@luotao1 We were able to reproduce the problem on AVX512 platform. @jczaja is investigating this

**Comment by jczaja** [comment]
@luotao1 
I investigated this issue and want to share results of investigation.

Problem is with computation of square root as a part of BatchNorm _(1 / sqrt(Var(x) + Eps))_ . PaddlePaddle/EIGEN is computing with **lower** precision than MKLDNN. In fact PaddlePaddle is using EIGEN which by default is having EIGEN_FAST_MATH set to 1. This flag set to one makes EIGEN to compute square root in a fast (inaccurate) way . Here is a relevant code from eigen: 
https://github.com/eigenteam/eigen-git-mirror/blob/767b34a01878cd7ddb566d5fafa3b414e1dd9d6f/Eigen/src/Core/arch/SSE/MathFunctions.h#L445

inacuracy in computing square root makes inv_std to differ within 10e-6 range within BatchNorm, but after whole inference error is does increase to exceed 10e-4.

To test this thesis, please add (at beginning) to your major CMakeLists.txt of Paddle:
`
add_definitions(-DEIGEN_FAST_MATH=0)`

This makes your test running on my platform.

Few Notes:
1) If you value EIGEN_FAST_MATH , then please adjust 
maximal inaccuracy margin for models with BatchNorm accordingly eg. starting from the point that result of square root has two bits diffrent from precise result.
2) Perhaps just checking the classification result (eg. if category of top-1 is as expected) could be good enough combined with less restricted error margin than 10e-4.

Please confirm if disabling EIGEN_FAST_MATH fixed your test

**Comment by luotao1** [comment]
@jczaja Thanks for your detail investigation. After `add_definitions(-DEIGEN_FAST_MATH=0)`, the test is fixed. We will see how to solve this issue totally next.

https://eigen.tuxfamily.org/dox/TopicPreprocessorDirectives.html
![image](https://user-images.githubusercontent.com/6836917/46592194-f4a2b380-caf2-11e8-9dd7-7d7df7eb1761.png)

**Comment by luotao1** [comment]
@jczaja Does MKLDNN have similar option like `EIGEN_FAST_MATH`, since we should keep the same accuracy now. There are 2 reasons:
- We use `use_fast_math` in cuda.cmake (see https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html for details)
https://github.com/PaddlePaddle/Paddle/blob/8cd17c04c1d0eadebb96c2a9dc887c8411e7e6c8/cmake/cuda.cmake#L178 
![image](https://user-images.githubusercontent.com/6836917/46592901-70076380-caf9-11e8-88e1-92f46e200feb.png)
![image](https://user-images.githubusercontent.com/6836917/46592916-92997c80-caf9-11e8-8914-1814e25286b7.png)
We should keep the accuracy same between GPU and CPU.
- A lot of models are trained using `EIGEN_FAST_MATH` and `use_fast_math`, if we disable them, the accuracy of a lot of pre-trained models may be affected.

Anyway, we will test disable`EIGEN_FAST_MATH` and `use_fast_math` at first to see whether https://github.com/PaddlePaddle/continuous_evaluation would be affected.

**Comment by luotao1** [comment]
reopen due to #13771
